### ISSUE TITLE
Horovod pytorch will hang on multi-node

### ISSUE THREAD

**Issue by GaofengCheng** [horovod/horovod#1086]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet): Pytorch
2. Framework version: 1.0.0
3. Horovod version: 0.16.1
4. MPI version: 4.0.1
5. CUDA version: 10.1
6. NCCL version: 2.4.6
7. Python version: 3.7
8. OS and version: centos 7.6

**Checklist:**
1. Did you search issues to find if somebody asked this question before? no
2. If your question is about hang, did you read [this doc] yes (https://github.com/horovod/horovod/blob/master/docs/running.md)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.md)?

**Your question:**
I have two nodes: rd06.rd.speech.lan and rd03.rd.speech.lan
I run `CUDA_VISIBLE_DEVICES=0,1,2,3 horovodrun -np 4 -H localhost:4 python pytorch_mnist.py ` separately on the two nodes, it's OK, and the NCCL is activated.

But the conmmand `CUDA_VISIBLE_DEVICES=0,1,2,3 horovodrun -np 8 -H rd03.rd.speech.lan:4,rd06.rd.speech.lan:4 python pytorch_mnist.py` , it will print

`[rd06.rd.speech.lan:59439] Warning: could not find environment variable "OLDPWD"`

and hang. The single node will also throw this warning but it didn't affect the training.

I interruppted the command and it returns:
```
^Zmpirun: Forwarding signal 20 to job
--------------------------------------------------------------------------
ORTE does not know how to route a message to the specified daemon
located on the indicated node:

  my node:   rd06
  target node:  rd03

This is usually an internal programming error that should be
reported to the developers. In the meantime, a workaround may
be to set the MCA param routed=direct on the command line or
in your environment. We apologize for the problem.
--------------------------------------------------------------------------
```

The ip config is as:
rd03.rd.speech.lan
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever

2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether e4:43:4b:26:0d:da brd ff:ff:ff:ff:ff:ff
    inet 10.20.7.203/24 brd 10.20.7.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::e643:4bff:fe26:dda/64 scope link 
       valid_lft forever preferred_lft forever

6: ib0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 2044 qdisc mq state UP group default qlen 256
    link/infiniband a0:00:02:20:fe:80:00:00:00:00:00:00:50:6b:4b:03:00:67:9a:a1 brd 00:ff:ff:ff:ff:12:40:1b:ff:ff:00:00:00:00:00:00:ff:ff:ff:ff
    inet 110.20.7.203/24 brd 110.20.7.255 scope global ib0
       valid_lft forever preferred_lft forever
    inet6 fe80::526b:4b03:67:9aa1/64 scope link 
       valid_lft forever preferred_lft forever
```
rd06.rd.speech.lan
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever

2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 6c:92:bf:b6:98:c6 brd ff:ff:ff:ff:ff:ff
    inet 10.20.7.206/24 brd 10.20.7.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::6e92:bfff:feb6:98c6/64 scope link 
       valid_lft forever preferred_lft forever

6: ib0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 2044 qdisc mq state UP group default qlen 256
    link/infiniband a0:00:02:20:fe:80:00:00:00:00:00:00:50:6b:4b:03:00:67:99:81 brd 00:ff:ff:ff:ff:12:40:1b:ff:ff:00:00:00:00:00:00:ff:ff:ff:ff
    inet 110.20.7.206/24 brd 110.20.7.255 scope global ib0
       valid_lft forever preferred_lft forever
    inet6 fe80::526b:4b03:67:9981/64 scope link 
       valid_lft forever preferred_lft forever
```

**Comment by alsrgv** [comment]
@GaofengCheng, can you try [this suggestion](https://github.com/horovod/horovod/issues/952#issuecomment-492874494) and see if it helps?

**Comment by GaofengCheng** [comment]
@alsrgv how to add `-mca routed direct` for horovodrun?

**Comment by alsrgv** [comment]
```bash
$ echo "routed = direct" >> /usr/local/etc/openmpi-mca-params.conf
```

**Comment by GaofengCheng** [comment]
No help.
```
CUDA_VISIBLE_DEVICES=0,1,2,3 horovodrun -np 8 -H rd03.rd.speech.lan:4,rd06.rd.speech.lan:4 --verbose python pytorch_mnist.py 
Filtering local host names.
Checking ssh on all remote hosts.
SSH was successful into all the remote hosts.
Testing interfaces on all the hosts.
Launched horovodrun server.
Attempted to launch horovod task servers.
Waiting for the hosts to acknowledge.
Notified all the hosts that the registration is complete.
Waiting for hosts to perform host-to-host interface checking.
Host-to-host interface checking successful.
Interfaces on all the hosts were successfully checked.
mpirun --allow-run-as-root --tag-output -np 8 -H rd03.rd.speech.lan:4,rd06.rd.speech.lan:4 -bind-to none -map-by slot -mca pml ob1 -mca btl ^openib  -mca btl_tcp_if_include eth0 -x NCCL_DEBUG=INFO -x NCCL_SOCKET_IFNAME=eth0 -x CUDA_VISIBLE_DEVICES -x CPLUS_INCLUDE_PATH -x CUDA_PATH -x OPEN_MPI -x XDG_SESSION_ID -x HOSTNAME -x SELINUX_ROLE_REQUESTED -x TERM -x SHELL -x HISTSIZE -x SSH_CLIENT -x CONDA_SHLVL -x LIBRARY_PATH -x CONDA_PROMPT_MODIFIER -x SELINUX_USE_CURRENT_RANGE -x OLDPWD -x SSH_TTY -x CUDA_HOME -x USER -x LD_LIBRARY_PATH -x LS_COLORS -x CONDA_EXE -x _CE_CONDA -x MAIL -x PATH -x CONDA_PREFIX -x C_INCLUDE_PATH -x PWD -x LANG -x SELINUX_LEVEL_REQUESTED -x _CE_M -x CXX -x HISTCONTROL -x KRB5CCNAME -x SHLVL -x HOME -x CONDA_PYTHON_EXE -x LOGNAME -x SSH_CONNECTION -x MPI_HOME -x CONDA_DEFAULT_ENV -x LESSOPEN -x CC -x XDG_RUNTIME_DIR -x _ -x _HOROVOD_SECRET_KEY python pytorch_mnist.py
[rd06.rd.speech.lan:67985] Warning: could not find environment variable "OLDPWD"
```

**Comment by GaofengCheng** [comment]
I have tried pytorch inbuilt distribution, it works between these two nodes.

**Comment by alsrgv** [comment]
Could you try this:
```
$ CUDA_VISIBLE_DEVICES=0,1,2,3 horovodrun -np 8 -H 10.20.7.203:4,10.20.7.206:4 --verbose python pytorch_mnist.py 
```

I've started #1087 to resolve the warning.  It's safe to ignore.

**Comment by GaofengCheng** [comment]
tried.
```
CUDA_VISIBLE_DEVICES=0,1,2,3 horovodrun -np 8 -H 10.20.7.203:4,10.20.7.206:4 --verbose python pytorch_mnist.py 
Filtering local host names.
Checking ssh on all remote hosts.
ssh not successful for host 10.20.7.203:
Permission denied, please try again.
Received disconnect from UNKNOWN port 65535:2: Too many authentication failures
Authentication failed.
```

**Comment by alsrgv** [comment]
@GaofengCheng, that's interesting.  Did you set up [passwordless auth](http://www.linuxproblem.org/art_9.html)?

**Comment by GaofengCheng** [comment]
the two nodes have both eth0 and ib0. but the horovod common_intfs choose eth0. How to force the horovod use ib0. If I run locally, nccl will choose ib0.

**Comment by alsrgv** [comment]
I suspect it has something to do with the routing table.  What’s the output of `ip route`?

`horovodrun` is a convenience wrapper. You can get full control by using `mpirun` directly - https://github.com/horovod/horovod/blob/master/docs/mpirun.rst.

**Comment by GaofengCheng** [comment]
ip route

rd03:
default via 10.20.7.1 dev eth0 
10.20.7.0/24 dev eth0 proto kernel scope link src 10.20.7.203 
10.20.7.0/24 dev ib0 proto kernel scope link src 10.20.7.223 
169.254.0.0/16 dev eth0 scope link metric 1002 
169.254.0.0/16 dev ib0 scope link metric 1006

rd06:
default via 10.20.7.1 dev eth0 
10.20.7.0/24 dev eth0 proto kernel scope link src 10.20.7.206 
10.20.7.0/24 dev ib0 proto kernel scope link src 10.20.7.226 
169.254.0.0/16 dev eth0 scope link metric 1002 
169.254.0.0/16 dev ib0 scope link metric 1006

**Comment by alsrgv** [comment]
@GaofengCheng, I think the issue is that traffic to `110.20.7.206/24` gets routed via `10.20.7.206/24` and thus ib0 is not used (in `horovodrun`) and also makes Open MPI confused.

If you're feeling adventurous and don't mind playing with the routing table, changing `10.20.7.0/24 dev ib0 proto kernel scope link src 10.20.7.226` to `110.20.7.0/24 dev ib0 proto kernel scope link src 110.20.7.206` (and comparable change on the other machine), Open MPI may start working.

**Comment by zarzen** [comment]
you can use `-mca btl_tcp_if_include 110.20.7.0/24` for `mpirun` to let all traffic go through `ib0`
full version:
```
# do not use hostname, because hostname might bind with eth0
mpirun -np 8 -H 110.20.7.206:4,110.20.7.203:4 \
    -bind-to none -map-by slot \
    -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH \
    -x NCCL_SOCKET_IFNAME=^lo,docker0 \
    -mca pml ob1 -mca btl ^openib \
    -mca btl_tcp_if_include 110.20.7.0/24 \
    python train.py
```

**Comment by ieggel** [comment]
**Here my solution using workers with different network interface names using horovodrun**

I'm using Horovod 0.19.5.
I have 2 workers with different network interface names:
worker1: eno1
worker2: eno3
Both workers are in the same subnet 192.168.1.0/24.

I was able to use horovodrun (no mpirun required) by passing  `--network-interface` and `--mpi-args` to horovod.

```bash
horovodrun \
   -np 3 \
   -H worker1:2,worker2:1 \
   --verbose \
   --network-interface 153.109.124.0/24 \
   --mpi-args="-x NCCL_SOCKET_IFNAME=eno" \
   python myscript.py
```
Short explanation:
I used the horovod arg `--network-interface` to pass my subnet. It seems mpi can work with that and although with that alone my script correctly launchd the processes on the workers, it started hanging with no GPU usage at all.  It seems NCCL itself does not understand the subnet notation and still needs a network interface name. In order to make it work I had to also pass `--mpi-args=-x NCCL_SOCKET_IFNAME=eno` which tells NCCL to only consider network interfaces starting with 'eno'. 

Hope this can prevent some headaches.
